designs/RFC - Network Setup Teardown Optimization.md

```diff
@@ -0,0 +1,381 @@
+  %4 = comp.alloc %0 : (!comp.execenv<ocl:0,(11)>) -> memref<16x32xf32, 11>
+  %5 = comp.alloc %0 : (!comp.execenv<ocl:0,(11)>) -> memref<8x16xf32, 11>
+  %k1 = comp.create_kernel %0 {kernel = @main_kernel::@main_kernel} : (!comp.execenv<ocl:0,(11)>) -> !comp.kernel<ocl, 11, 8x32xf32>
+  %k2 = comp.create_kernel %0 {kernel = @main_kernel_0::@main_kernel} : (!comp.execenv<ocl:0,(11)>) -> !comp.kernel<ocl, 11, 16x32xf32, 8x16xf32, 8x32xf32>
+  comp.loop(%arg1: memref<8x16xf32>, %arg2: memref<16x32xf32>, %arg3: memref<8x32xf32>) {
+    %w1 = comp.schedule_write %arg3 to %1 on %0 : (!comp.execenv<ocl:0,(11)>, memref<8x32xf32>, memref<8x32xf32, 11>) -> !comp.event<ocl>
+    %2 = comp.launch_kernel %k1, %c4, %c1, %c1, %c1, %c2, %c32, %1 wait for %w1: (!comp.kernel<ocl, 11, 8x32xf32>, index, index, index, index, index, index, memref<8x32xf32, 11>, !comp.event<ocl>) -> !comp.event<ocl>
+    %w4 = comp.schedule_write %arg2 to %4 on %0 : (!comp.execenv<ocl:0,(11)>, memref<16x32xf32>, memref<16x32xf32, 11>) -> !comp.event<ocl>
+    %w5 = comp.schedule_write %arg1 to %5 on %0 : (!comp.execenv<ocl:0,(11)>, memref<8x16xf32>, memref<8x16xf32, 11>) -> !comp.event<ocl>
+    %7 = comp.launch_kernel %k2, %c1, %c1, %c1, %c8, %c4, %c1, %4, %5, %1 wait for %2, %w4, %w5: (!comp.kernel<ocl, 11, 16x32xf32, 8x16xf32, 8x32xf32>, index, index, index, index, index, index, memref<16x32xf32, 11>, memref<8x16xf32, 11>, memref<8x32xf32, 11>, !comp.event<ocl>, !comp.event<ocl>, !comp.event<ocl>) -> !comp.event<ocl>
+    %10 = comp.schedule_read %arg3 from %1 on %0 wait for %7 : (memref<8x32xf32>, memref<8x32xf32, 11>, !comp.execenv<ocl:0,(11)>, !comp.event<ocl>) -> !comp.event<ocl>
+    comp.wait %10 : !comp.event<ocl>
+  }
+  comp.destroy_kernel %k1 : !comp.kernel<ocl, 11, 8x32xf32>
+  comp.destroy_kernel %k2 : !comp.kernel<ocl, 11, 16x32xf32, 8x16xf32, 8x32xf32>
+  comp.dealloc %0 %4 : (!comp.execenv<ocl:0,(11)>, memref<16x32xf32, 11>) -> ()
+  comp.dealloc %0 %5 : (!comp.execenv<ocl:0,(11)>, memref<8x16xf32, 11>) -> ()
+  comp.dealloc %0 %1 : (!comp.execenv<ocl:0,(11)>, memref<8x32xf32, 11>) -> ()
+  comp.destroy_execenv %0 : !comp.execenv<ocl:0,(11)>
+  return
+}
+```
+
+### Final Lowering
+
+With a minor tweak, we could run the above code as a coroutine &mdash; we could pass in an extra parameter to represent
+a bidirectional request/response pipe, and pass it to `comp.loop`.  Our sense is that it's more obvious to split the
+current `main()` function into multiple functions invoked individually to set up the network, run it, and tear it down.
+
+To do this, we add a new `!comp.network` to serve as a resource container, and add construction and deconstruction
+operations for it; this allows us to return a single value to the caller to represent the constructed network, and in
+the lowering to LLVMIR, we can trivially replace it with a pointer to a heap-allocated well-typed struct.  For our
+example code, this transformation looks something like this:
```

Comment:
What's the advantage of `!comp.network`? I think we'll find that it's often easier to pass each value standalone, not contained within a 'struct'.

Reply:
I'm thinking the thing doing the passing is `JitExecutable`; in its constructor, it's running `plaidml_init`, getting some value back to hold the network, and then using that value for its `plaidml_exec` calls and then passing it to `plaidml_fini` in its destructor.  It can't directly handle a variable number of return values and arguments.  So something, somewhere, on the JITted code side, needs to be boxing the variables being passed from init to exec and fini.  Given that, I think it's simplest to have an explicit box at the comp dialect, and lower it to a heap-allocated LLVMIR struct.

I'm definitely open to other ideas, though -- maybe there's some MLIR facility for handling this sort of thing?  Some other good pattern people are using?

Reply:
@jbruestle and I just had a long chat about ways to improve this: we think we've found a way to get rid of the !comp.network type for passing values between the init and exec/fini functions, without running into all the complexity I was seeing in the lowering and execution.  I'm working on a revision.

Reply:
PTAL?